File: 57-insert-interval/solution.ts
function insert(intervals: number[][], newInterval: number[]): number[][] {
  let now: number[] = [];
  const answer: number[][] = [];
  let len = intervals.length;
  let flag = false;

  if (newInterval[0] >= (intervals[len - 1]?.[0] ?? -1)) {
    intervals.push(newInterval);
    len += 1;
  }

  for (let i = 0; i < len; i++) {
    let [start, end] = intervals[i];
    if (newInterval[0] < start && flag === false) {
      start = newInterval[0];
      end = newInterval[1];
      flag = true;
      i--;
    }

    if (now.length === 0) now = [start, end];
    else if (start <= now[1]) {
      now = [now[0], Math.max(end, now[1])];
    } else {
      answer.push(now);
      now = [start, end];
    }
  }
  answer.push(now);

  return answer;
}

File: 57-insert-interval/solution.cpp
class Solution {
public:
    bool isOverlap(vector<int>& int1, vector<int>& int2) {
        if(int1[1] < int2[0] || int1[0] > int2[1]) return false;
        return true;
    }

    vector<vector<int>> insert(vector<vector<int>>& intervals, vector<int>& newInterval) {
        if(intervals.size() == 0) return {newInterval};
        vector<vector<int>> ans;
        bool flag = false;
        int left = INT_MAX, right = INT_MIN;
        int n = intervals.size();
        if(newInterval[1] < intervals[0][0]) ans.push_back(newInterval);

        for(int i = 0; i < n; i++) {
            if(i > 0 && newInterval[0] > intervals[i - 1][1] && newInterval[1] < intervals[i][0]) ans.push_back(newInterval);

            if(isOverlap(intervals[i], newInterval)) {
                flag = true;
                left = min(left, min(intervals[i][0], newInterval[0]));
                right = max(right, max(intervals[i][1], newInterval[1]));
            } else {
                if(flag) {
                    ans.push_back({left, right});
                    flag = false;
                }
                ans.push_back(intervals[i]);
            }
        }
        if(flag) ans.push_back({left, right});
        if(newInterval[0] > intervals[n - 1][1]) ans.push_back(newInterval);
        return ans;
    }
};
